### DISCUSSION TITLE
Help with solving microphone feedback by delaying the I2S stream

### DISCUSSION THREAD

**Discussion by cmcooper1971** [pschatzmann/arduino-audio-tools#1708] [Q&A] [answered]
Hi All,

A small hiccup preventing launch at https://nor-con.co.uk/ !

When using the I2S Microphone in full costume (see pics at end of thread), the plastic panels of the suits cause feedback.

Reading online, one way to reduce the feedback is to create a delay between the I2S input and the I2S output. 

I am out of time and haven't the time to learn this part, so anyone who can review this code and update accordingly, it would be sincerely appreciated.

Microphone Function

```
// Enable MIC & disable Player

void enableMIC() {

	// Stop player

	decoderWav.end();
	decoderMP3.end();
	mp3.end();
	in.end();
	out.end();

	// Start I2S in

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);
	debugln("Starting I2S...");

	auto config_in = in.defaultConfig(RX_MODE);
	config_in.copyFrom(infoMic);
	config_in.i2s_format = I2S_STD_FORMAT;
	config_in.is_master = true;
	config_in.port_no = 0;
	config_in.pin_bck = I2S_BCLK_0;
	config_in.pin_ws = I2S_LRC_0;
	config_in.pin_data = I2S_DIN_0;
	in.begin(config_in);

	// Start I2S out

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(infoMic);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	// Start the VolumeMeter

	vMeter.begin(infoMic);

	// Start the microphone stream

	mic.begin(out, converter);

	debugln("Mic Enabled");
	digitalWrite(amberLED, LOW);
	playerControlFlagChanged = false;

	displayVolumeRegister(false);

} // Close function
```

Loop trigger to play I2S microphone input:

```
if (micControlFlag == true && playerControlFlag == false) {

	mic.copy();

	// Get the max amplitude using VolumeMeter

	int volume = vMeter.volume();
	//debug("Volume: ");				// Do not print unless recalibration is needed
	//debugln(volume);					// Do not print unless recalibration is needed


	if (volume > volumeThreshold) {		// Adjust this threshold as necessary

		volumeSampleCount++;

		if (volumeSampleCount > minVolumeDuration) {

			isSoundDetected = true;
			volumeSampleCount = 0;

			onSoundStart();				// Trigger event when sound is detected

		}
	}

	else {

		if (isSoundDetected) {

			silenceSampleCount++;

			if (silenceSampleCount > minSilenceDuration) {

				isSoundDetected = false;
				silenceSampleCount = 0;

				onSoundStop();  // Trigger event when sound stops

			}
		}
	}
}
```

This all works a treat, I somehow need to create a delay which I can vary to get the best results in reducing the feedback.

Thx.

![IMG_6914](https://github.com/user-attachments/assets/63772360-ae55-4c45-8d99-a48b7238e24e)
![IMG_6915](https://github.com/user-attachments/assets/46299a99-edab-456c-9ff8-b7efed298446)

**Comment by fobi55** [comment]
I'm afraid that the delay must be adaptive. This is a complicated task. Probably using unidirectional microphone could be a fast solution in your case.
BR

**Comment by pschatzmann** [comment, marked as answer]
Based on the answer of fobi55, I would suggest build some kind of waveguide both around the microphone and the speakers to minimize the chance that the microphone picks up the sound from the speakers.

In addition, decrease the overall amplification or you can try to use the Equilizer to decrease just the loudest frequencies.

**Comment by cmcooper1971** [comment]
Thanks Robi55 and Phil, 

Ill try the enclosure idea as that is the easiest to implement.

Your replies are appreciated.
